### PR TITLE
Initial version of retrieve_tool_func_result to support generate_index_json, reverse_generate_index_json

### PR DESCRIPTION
**Pull request by liuyuhang13**
# Description

Implementing `retrieve_tool_func_result` to support `generate_index_json`, `reverse_generate_index_json`, so that tool input can be generated by  call a func and use other inputs as func inputs.

# All Promptflow Contribution checklist:
- [ x ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ x ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: examples/tools/tool-package-quickstart/my_tool_package/yamls/tool_with_generate_index.yaml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+          type:
+          - string
+          reference: ${inputs.content_field}
+        - embedding_deployment:
+          type:
+          - string
+          optional: true
+          reference: ${inputs.embedding_deployment}
+        reverse_func_path: my_tool_package.tools.tool_with_generate_index.reverse_generate_index_json
+      input_type: UIHide
```
Comment (dorisjoy) [inline examples/tools/tool-package-quickstart/my_tool_package/yamls/tool_with_generate_index.yaml:28]: delete this input_type

File: examples/tools/tool-package-quickstart/my_tool_package/yamls/tool_with_generate_index.yaml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    index_type:
+      type: 
+      - string
+      enum:
+        - Azure Cognitive Search
+        - MLIndex
+      input_type: UIOnly
```
Comment (dorisjoy) [inline examples/tools/tool-package-quickstart/my_tool_package/yamls/tool_with_generate_index.yaml:35]: uionly_hidden

File: examples/tools/tool-package-quickstart/my_tool_package/yamls/tool_with_generate_index.yaml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      dynamic_list:
+        func_path: my_tool_package.tools.tool_with_generate_index.list_indexes
+        func_kwargs: 
+        - name: index_type
+          type: 
+          - string
+          optional: true
+          reference: ${inputs.index_json.index_type}
+      allow_manual_entry: false
+      is_multi_select: false
```
Comment (dorisjoy) [inline examples/tools/tool-package-quickstart/my_tool_package/yamls/tool_with_generate_index.yaml:50]: Need to add input_type for index_assetId / content_field / embedding_deployment

File: examples/tools/tool-package-quickstart/my_tool_package/tools/tool_with_generate_index.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
+from promptflow import tool
+from typing import Dict, List
```
Comment (dorisjoy) [inline examples/tools/tool-package-quickstart/my_tool_package/tools/tool_with_generate_index.py:2]: Change tool name to tool_with_generated_by_input

File: examples/tools/tool-package-quickstart/my_tool_package/yamls/tool_with_generate_index.yaml [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      dynamic_list:
+        func_path: my_tool_package.tools.tool_with_generate_index.list_embedding_deployment
+        func_kwargs: 
+        - name: index_type
+          type: 
+          - string
+          optional: true
+          reference: ${inputs.index_json.index_type}
+  module: my_tool_package.tools.tool_with_generate_index
+  name: MLIndex Tool
```
Comment (dorisjoy) [inline examples/tools/tool-package-quickstart/my_tool_package/yamls/tool_with_generate_index.yaml:78]: name: Tool with Generated By Input

File: examples/tools/tool-package-quickstart/my_tool_package/yamls/tool_with_generate_index.yaml [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  description: This is my tool combined dynamic list and enabled by
```
Comment (dorisjoy) [inline examples/tools/tool-package-quickstart/my_tool_package/yamls/tool_with_generate_index.yaml:79]: description: This is a tool with generated by input

File: src/promptflow/promptflow/_sdk/_utils.py [hunk 7]
```diff
@@ -624,6 +624,26 @@ def _generate_tool_meta(
...
+    func_kwargs = function_config.get("func_kwargs", {})
+
+    from promptflow._cli._utils import get_workspace_triad_from_local
+
+    workspace_triad = get_workspace_triad_from_local()
+    if workspace_triad.subscription_id and workspace_triad.resource_group_name and workspace_triad.workspace_name:
+        return retrieve_tool_func_result(func_call_scenario, func_path, func_kwargs, workspace_triad._asdict())
+    # if no workspace triple available, just skip.
+    else:
+        return retrieve_tool_func_result(func_call_scenario, func_path, func_kwargs)
```
Comment (dorisjoy) [inline src/promptflow/promptflow/_sdk/_utils.py:644]: return {"result": result, "logs": {}}

File: examples/tools/tool-package-quickstart/my_tool_package/yamls/tool_with_generate_index.yaml (hunk 5 above)
Comment (liuyuhang13) [reply examples/tools/tool-package-quickstart/my_tool_package/yamls/tool_with_generate_index.yaml]: updated

File: src/promptflow/promptflow/_sdk/_utils.py (hunk 7 above)
Comment (liuyuhang13) [reply src/promptflow/promptflow/_sdk/_utils.py:644]: updated

**Comment by chjinche** [comment]
fix typo `retrieve` in title and description.

File: src/promptflow/promptflow/_sdk/_utils.py [hunk 8]
```diff
@@ -624,6 +624,31 @@ def _generate_tool_meta(
...
+def _retrieve_tool_func_result(func_call_scenario: str, function_config: Dict):
+    """Retrieve tool func result according to func_call_scenario.
+
+    :param func_call_scenario: function call scenario
+    :param function_config: function config in tool meta. Should contain'func_path' and 'func_kwargs'.
+    :return: func call result according to func_call_scenario.
+    """
+    func_path = function_config.get("func_path", "")
+    func_kwargs = function_config.get("func_kwargs", {})
+
```
Comment (chjinche) [inline src/promptflow/promptflow/_sdk/_utils.py:636]: # May call azure control plane api in the custom function to list Azure resources.
    # which may need Azure workspace triple.
    # TODO: move this method to a common place.

Comment (liuyuhang13) [reply src/promptflow/promptflow/_sdk/_utils.py]: updated.

**Review by dorisjoy — APPROVED** [review]

**Review by chjinche — APPROVED** [review]

**Pull request merged**